### PR TITLE
FCBHDBP-437 DBP - Enable "Discover search" and "search My Bible" to find scripture references such as "Romans 2:4"

### PR DESCRIPTION
**Pull request by vichugofsl**
# Description
So, the goal is about building two new endpoint to fulfill the request to search scripture references from global search and scripture references from “Search My bible“.

1. It would be an endpoint to fetch the bible_verses by Language Code (ISO)/Book ID/Chapter/Verse but the Verse could be optional. This endpoint to fulfill the request to search scripture references from the global search. We can know the language Code because the user that has logged in app, has selected a default language and it will be a good filter to avoid that the SQL query is slow. The route would be something like: `/bibles/verses/{language_code}/{book_id}/{chapter_id}/{verse_number?}`

2. The second one endpoint will be to fulfill the request to search scripture references from “Search My bible“. This endpoint would be similar to the above endpoint but, it will use the bible_id rather language_id because the app is able to know the bible_id in this part. The route would be something like: `/bible/{bible_id}/verses/{book_id}/{chapter_id}/{verse_number?}`

## Issue Link
https://fullstacklabs.atlassian.net/browse/FCBHDBP-437

## How Do I QA This
- I create a Postman folder with 4 tests.
https://digitalbibleplatform.postman.co/workspace/DBP4-workspace~0a283f7f-06d9-467d-97f0-fef92d18bd3d/folder/12519377-47ae182e-c9fa-4fac-a11b-3dfa13519d7c

### CODE REVIEW

File: app/Http/Controllers/Bible/BibleVersesController.php [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+     * Display a listing of the bibles.
+     *
+     * @OA\Get(
+     *     path="/bibles/verses/{language_id}/{book_id}/{chapter_id}/{verse_number?}",
+     *     tags={"BibleVerses"},
+     *     summary="Returns Bibles Verses based on filter criteria",
+     *     description="The base bible route returning by default bibles and filesets that your key has access to",
+     *     operationId="v4_bible_verses.verse_by_language",
+     *     @OA\Parameter(
+     *          name="language_id",
```
Comment (bradflood) [inline app/Http/Controllers/Bible/BibleVersesController.php:29]: language should be required

File: app/Http/Controllers/Bible/BibleVersesController.php [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+     *          in="path",
+     *          @OA\Schema(ref="#/components/schemas/Language/properties/id"),
+     *          description="",
+     *     ),
+     *     @OA\Parameter(
+     *          name="book_id",
```
Comment (bradflood) [inline app/Http/Controllers/Bible/BibleVersesController.php:35]: book should be required

File: app/Http/Controllers/Bible/BibleVersesController.php [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+     * Display a listing of the bibles.
+     *
+     * @OA\Get(
+     *     path="/bible/{bible_id}/verses/{book_id}/{chapter_id}/{verse_number?}",
+     *     tags={"BibleVerses"},
+     *     summary="Returns Bibles Verses based on filter criteria",
+     *     description="The base bible route returning by default bibles and filesets that your key has access to",
+     *     operationId="v4_bible_verses.verse_by_bible",
+     *     @OA\Parameter(
+     *          name="bible_id",
```
Comment (bradflood) [inline app/Http/Controllers/Bible/BibleVersesController.php:104]: book should be required

**Review by bradflood — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/Bible/BibleVersesController.php [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+     *          in="path",
+     *          @OA\Schema(ref="#/components/schemas/Bible/properties/id"),
+     *          description="",
+     *     ),
+     *     @OA\Parameter(
+     *          name="book_id",
```
Comment (bradflood) [inline app/Http/Controllers/Bible/BibleVersesController.php:110]: book should be required

**Review by bradflood — CHANGES_REQUESTED** [review]
can you give some timing info on these endpoints?

File: app/Http/Controllers/Bible/BibleVersesController.php (hunk 4 above)
Comment (vichugofsl) [reply app/Http/Controllers/Bible/BibleVersesController.php:114]: It has been updated and it is a required parameter.

File: app/Http/Controllers/Bible/BibleVersesController.php (hunk 3 above)
Comment (vichugofsl) [reply app/Http/Controllers/Bible/BibleVersesController.php:107]: It has been updated and it is a required parameter.

File: app/Http/Controllers/Bible/BibleVersesController.php (hunk 2 above)
Comment (vichugofsl) [reply app/Http/Controllers/Bible/BibleVersesController.php:36]: It is required parameter now.

File: app/Http/Controllers/Bible/BibleVersesController.php (hunk 1 above)
Comment (vichugofsl) [reply app/Http/Controllers/Bible/BibleVersesController.php]: It is required parameter now.

**Comment by vichugofsl** [comment]
> can you give some timing info on these endpoints?

@bradflood , I have tested both endpoints on my local environment and the time response for both endpoint is less than 100 ms.

Screenshot from 2022-09-23 17-15-33

File: app/Http/Controllers/Bible/BibleVersesController.php (hunk 3 above)
Comment (vichugofsl) [reply app/Http/Controllers/Bible/BibleVersesController.php:107]: @bradflood My bad, I missed adding more information, I have added the constraint to validate that some parameters should be required into the routing. In the following line you can see that I have used the methods: `whereAlphaNumeric` or `whereNumber` to validate the parameters through the routing. According to the above methods the parameters should pass the validations to go the controller.

https://github.com/faithcomesbyhearing/dbp/pull/838/files#diff-0e4323e6a03c91c9c8dc88afef15fdad09636468c0ac3550f497618372788b6dR130

```php
Route::name('v4_bible_verses.verse_by_language')->get(
    '/bibles/verses/{language_code}/{book_id}/{chapter_id}/{verse_number?}',
    'Bible\BibleVersesController@showVerseByLanguage'
)->whereAlphaNumeric('language_code')
->whereAlphaNumeric('book_id')
->whereNumber('chapter_id')
->whereNumber('verse_number');

Route::name('v4_bible_verses.verse_by_bible')->get(
    '/bible/{bible_id}/verses/{book_id}/{chapter_id}/{verse_number?}',
    'Bible\BibleVersesController@showVerseByBible'
)->whereAlphaNumeric('bible_id')
->whereAlphaNumeric('book_id')
->whereNumber('chapter_id')
->whereNumber('verse_number');
```

**Pull request merged**